### DISCUSSION TITLE
corfu and ESS: accessing list elements without having first to select list

### DISCUSSION THREAD

**Discussion by maxecharel** [minad/corfu#451] [Q&A]
Hi @minad,

I've tried `corfu` with `eglot`, works pretty well (amazing work, as always). 
I decided to give it a try with `ESS`, which relies on CAPF + some Company-specific backends (so I also took that as an opportunity to try `cape`; also a really cool package).

In R, dataframes are named lists, and their elements (i.e. features/columns/variables) can be accessed using the `$` operator.
In the example below, I consider `iris`, a builtin dataset, but the result is the same if I assign it to `foo` or `bar`. 

With Company, I can type (in one shot) `iris$`, and the popup will suggests the different elements it contains:
![acess_list_elements_ess_company](https://github.com/minad/corfu/assets/61969138/6b288079-1ccb-46ba-be91-f366c20e1934)

However, I failed to have Corfu behaving the same way. If I type `iris$` in one shot, Corfu will only suggest the corresponding object candidate, that is, `iris`. I then have to select it, and then use the `$` operator to have Corfu suggesting the different elements of the list:
![access_list_elements_ess_corfu](https://github.com/minad/corfu/assets/61969138/894220ce-979e-4212-87dc-981458170001)
 
I tried to play with `corfu-preselect`, but didn't succeed to get the desired behavior. Same when I tried to change the order the different backends were added to `completion-at-point-functions` via `cape-company-to-capf`.

Is there a way to make `corfu` behave the way `company` does? Could my problem be due to the fact that `corfu` treats as `$` as the regex metacharacter? 

Thanks a lot for the great work!

Best,
M

Emacs 29.3
Corfu 20240409.438 (Melpa)
Cape 20240411.2207 (Melpa)

**Comment by maxecharel** [comment]
The more I think about it, the more I think the problem is due to `$` being treated as the 'end of string' metacharacter; also note that I use `orderless`

**Reply by minad** [reply]
Yes, likely.

**Reply by maxecharel** [reply]
I will have a look at how to prevent `corfu`/`orderless` to treat it as such; I'll keep you posted; thanks

**Reply by maxecharel** [reply]
(NB: as expected disabling `orderless` 'solves' the problem; next step is to keep `orderless` enabled and make it treat `$` as a standard character)

**Comment by maxecharel** [comment]
Hi,
I ended up implementing the following workaround:
```elisp
(defun mpger/cape-capf-ess ()
  (dolist (elmnt
	   (mapcar #'cape-company-to-capf
		   (list #'company-R-library #'company-R-args #'company-R-objects)))
    (add-to-list 'completion-at-point-functions elmnt))
  (setq-local orderless-matching-styles '(orderless-literal))) ;; default is '(orderless-literal orderless-regexp)
(use-package cape
  :hook
  (ess-r-mode . mpger/cape-capf-ess))
```

So, at the end of the day, to get the intended behavior I had to locally disable `orderless-regexp` matching style.
However, this is unfortunate because this is a pretty cool functionality of `orderless`. @minad  would it be worth adding a feature request to `orderless`, something like a custom variable allowing to exclude `$` (or other characters) from the metacharacters  in order to keep `orderless-regexp` as a matching style?

I guess that would be a substantial amount of work for a corner case, this is the reason I prefer to ask directly here instead of 'polluting' the issue tracker of `orderless`.

**Reply by maxecharel** [reply]
(Note to self) problems seems to come from the combination orderless + company backends via cape: latter not enabled in `inferior-ess-r-mode`, while matching styles include `orderless-regexp`, and I have no problem having the list elements suggested by Corfu after having typed e.g. `df$` (and without preselecting it) --> to be investigated

**Comment by minad** [comment]
Max Pger ***@***.***> writes:

> So, at the end of the day, to get the intended behavior I had to locally disable `orderless-regexp` matching style.
> However, this is unfortunate because this is a pretty cool functionality of
> `orderless`. @minad would it be worth to add a feature request to `orderless`,
> like a custom variable allowing to exclude `$` from the metacharacters in order
> to keep `orderless-regexp` as a matching style?

I don't think it would be worth it. But you can probably already achieve
your desired result with a custom Orderless matching style or style
dispatcher.

**Reply by maxecharel** [reply]
Ok thanks. I will try to investigate the Orderless customization you suggest. If you're ok I'll let this discussion open. If I succeed in achieving the desired result I'll post the solution here.

**Comment by kaibagley** [comment]
Hey @maxecharel, I'm having some other issues with Corfu+ESS. You mentioned above that you used `cape`, and I'm wondering how you set it up since it may work around the issue I'm having https://github.com/minad/corfu/discussions/387.

Are you just using `cape-company-to-capf` on the ESS built in Company backends?

Or are you using something like `eglot` to get around this issue? (I'm also having time-out issues getting eglot to run the r languageserver)

Thanks!
